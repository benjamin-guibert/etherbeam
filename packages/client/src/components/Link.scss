@use '../styles/resources/colors';
@use '../styles/resources/sizes';

.my-link,
.my-link:visited {
  cursor: pointer;
  color: colors.$secondary;
  text-decoration: inherit;
  padding: sizes.$space-xs;
}

.my-link:hover {
  filter: brightness(colors.$lighter);
}

.my-link:focus,
.my-clickabletr:focus {
  outline: 0px;
  box-shadow: 0px 0px 3px lighten(colors.$primary, percentage(colors.$variant-1));
}
