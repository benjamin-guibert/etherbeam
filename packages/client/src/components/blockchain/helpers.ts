import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from '../../libraries/ethereum/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}
