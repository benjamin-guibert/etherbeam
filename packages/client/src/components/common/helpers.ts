import { faDotCircle, faFile, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/ethereum/types'
import { BigNumber, ethers } from 'ethers'

const ETH_SYMBOL = 'ETH'
const ETH_DECIMALS = 18

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
  withSymbol?: boolean
}

const printAmount = (amount: string, symbol: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals, withSymbol } = options ?? {}
  const decimalsCount = forcedDecimals ?? decimals
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`
  const printedAmount = amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'

  return `${withSymbol ? `${symbol} ` : ''}${printedAmount}`
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_SYMBOL, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { symbol, decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, symbol, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faDotCircle
    default:
      return null
  }
}
