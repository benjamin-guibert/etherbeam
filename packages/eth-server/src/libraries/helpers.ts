import { Contract, ContractType, Transaction, USDC_HASH, WETH_HASH } from './types'
import uniq from 'lodash/uniq'

export const sleep = (milliseconds: number, callback?: () => void): Promise<void> => {
  return new Promise((resolve) => setTimeout(callback || resolve, milliseconds))
}

export const tryAction = async <T>(
  action: () => Promise<T>,
  waitingTime: number,
  maxRetries: number,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onError?: (error: any, count: number) => boolean | void,
  onNull?: (count: number) => boolean | void
): Promise<T> => {
  let result = undefined
  let count = 0

  while (!result) {
    try {
      result = await action()

      if (result == null) {
        count++
        if (onNull?.(count) || count > maxRetries) return null

        await sleep(waitingTime)
      }
    } catch (error) {
      count++
      if (onError?.(error, count) || count > maxRetries) return null

      await sleep(waitingTime)
    }
  }

  return result
}

export const sanitizeHash = (value: string): string => {
  if (!value) return value

  return `${value.substr(0, 2)}${value.substr(2).toLowerCase()}`
}

export const getAddressHashesFromTransaction = ({ from, to, action }: Transaction): string[] => {
  const hashes = [sanitizeHash(from), sanitizeHash(to)]
  action?.parameters.forEach((parameter) => {
    switch (parameter.type) {
      case 'address':
        hashes.push(sanitizeHash(parameter.value as string))
        break
      case 'address[]':
        ;(parameter.value as string[]).forEach((value) => hashes.push(sanitizeHash(value)))
        break
    }
  })

  return uniq(hashes)
}

export const isTransactionWatched = (transaction: Transaction, contracts: Contract[]): boolean => {
  const ignoredHashes = [WETH_HASH, USDC_HASH]
  const tokenHashes = contracts
    .filter((contract) => contract.type == ContractType.Token && !ignoredHashes.includes(contract.hash))
    .map((token) => token.sanitizedHash)
  const transactionHashes = getAddressHashesFromTransaction(transaction)

  return transactionHashes.some((hash) => tokenHashes.includes(hash))
}
