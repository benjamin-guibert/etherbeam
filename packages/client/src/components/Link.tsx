import React, { ReactElement, ReactNode } from 'react'
import './Link.scss'

interface LinkProps {
  title?: string
  href?: string
  onClick?: (event: React.SyntheticEvent) => void
  blank?: boolean
  active?: boolean
  disabled?: boolean
  children?: ReactNode
}

const Link = ({ title, href, onClick, blank, active, disabled, children }: LinkProps): ReactElement => {
  const onClickAction = (event: React.SyntheticEvent): void => {
    if (disabled) return

    event.stopPropagation()
    onClick?.(event)
  }

  const onKeyUp = (event: React.KeyboardEvent<HTMLAnchorElement>): void => {
    if (event.key == 'Enter') onClickAction(event)
  }

  const getClassName = (): string => {
    const classNames = ['my-link']
    if (disabled) classNames.push('my-disabled')
    if (active) classNames.push('my-active')

    return classNames.join(' ')
  }

  return (
    <a
      title={title}
      className={getClassName()}
      target={blank ? '_blank' : null}
      href={href && !disabled ? href : null}
      onClick={onClickAction}
      onKeyUp={onKeyUp}
      tabIndex={!disabled ? 0 : null}
    >
      {children}
    </a>
  )
}

export default Link
