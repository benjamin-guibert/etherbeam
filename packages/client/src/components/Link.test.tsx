import React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { clearMocks } from '../../tests/helpers'
import Link from './Link'

describe('<Link />', () => {
  const onClick = jest.fn()

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" href="/href" onClick={onClick}>
          Link
        </Link>
      )
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('title')).toBe('Title')
      expect(a.prop('href')).toBe('/href')
      expect(a.prop('target')).toBe(null)
      expect(a.prop('tabIndex')).toBe(0)
      expect(a.prop('children')).toBe('Link')
      expect(a.hasClass('my-link')).toBeTruthy()
      expect(a.hasClass('my-active')).toBeFalsy()
      expect(a.hasClass('my-disabled')).toBeFalsy()
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('On click', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" onClick={onClick}>
          Link
        </Link>
      )

      act(() => {
        component.find('a').simulate('click')
        component.update()
      })
    })

    it('should call onClick', () => expect(onClick).toBeCalledTimes(1))
  })

  describe('On key up (enter)', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" onClick={onClick}>
          Link
        </Link>
      )

      act(() => {
        component.find('a').simulate('keyUp', { key: 'Enter' })
        component.update()
      })
    })

    it('should call onClick', () => expect(onClick).toBeCalledTimes(1))
  })

  describe('On key up (other)', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" onClick={onClick}>
          Link
        </Link>
      )

      act(() => {
        component.find('a').simulate('keyUp', { key: 'a' })
        component.update()
      })
    })

    it('should not call onClick', () => expect(onClick).toBeCalledTimes(0))
  })

  describe('Blank', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" href="/href" onClick={onClick} blank>
          Link
        </Link>
      )
    })

    it('should render', () => expect(component.find('a').prop('target')).toBe('_blank'))
  })

  describe('Active', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" href="/href" onClick={onClick} active>
          Link
        </Link>
      )
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('tabIndex')).toBe(0)
      expect(a.hasClass('my-link')).toBeTruthy()
      expect(a.hasClass('my-active')).toBeTruthy()
      expect(a.hasClass('my-disabled')).toBeFalsy()
    })
  })

  describe('Disabled', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" href="/href" onClick={onClick} disabled>
          Link
        </Link>
      )
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('tabIndex')).toBeNull()
      expect(a.hasClass('my-link')).toBeTruthy()
      expect(a.hasClass('my-active')).toBeFalsy()
      expect(a.hasClass('my-disabled')).toBeTruthy()
    })
  })

  describe('Active & disabled', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Link title="Title" href="/href" onClick={onClick} active disabled>
          Link
        </Link>
      )
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('tabIndex')).toBeNull()
      expect(a.hasClass('my-link')).toBeTruthy()
      expect(a.hasClass('my-active')).toBeTruthy()
      expect(a.hasClass('my-disabled')).toBeTruthy()
    })
  })
})
