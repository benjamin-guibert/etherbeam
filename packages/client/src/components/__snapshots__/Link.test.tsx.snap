// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> Default should match snapshot 1`] = `
<a
  class="my-link"
  href="/href"
  tabindex="0"
  title="Title"
>
  Link
</a>
`;
